crates/ty_ide/src/docstring.rs

```diff
@@ -153,6 +149,128 @@ fn documentation_trim(docs: &str) -> String {
+/// line-breaks and indentation. Some constructs like *italic*, **bold**, and `inline code`
+/// are shared between the two formats. Other things like _italic_ are only markdown
+/// and should be escaped.
+fn render_markdown(docstring: &str) -> String {
+    // TODO: there is a convention that `singletick` is for items that can
+    // be looked up in-scope while ``multitick`` is for opaque inline code.
+    // While rendering this we should make note of all the `singletick` locations
+    // and (possibly in a higher up piece of logic) try to resolve the names for
+    // cross-linking. (Similar to `TypeDetails` in the type formatting code.)
+    let mut output = String::new();
+    let mut first_line = true;
+    let mut block_indent = 0;
+    let mut in_doctest = false;
+    let mut starting_literal = false;
+    let mut in_literal = false;
+    let mut in_any_code = false;
+    for untrimmed_line in docstring.lines() {
+        // We can assume leading whitespace has been normalized
+        let mut line = untrimmed_line.trim_start_matches(' ');
+        let line_indent = untrimmed_line.len() - line.len();
+
+        // First thing's first, add a newline to start the new line
+        if !first_line {
+            // If we're not in a codeblock, add trailing space to the line to authentically wrap it
+            if !in_any_code {
+                output.push_str("  ");
+            }
+            // Only push newlines if we're not scanning for a real line
+            if !starting_literal {
+                output.push('\n');
+            }
+        }
+
+        // If we're in a literal block and we find a non-empty dedented line, end the block
+        if in_literal && line_indent < block_indent && !line.is_empty() {
+            in_literal = false;
+            in_any_code = false;
+            block_indent = 0;
+            output.push_str("```\n");
+        }
+
+        // We previously entered a literal block and we just found our first non-blank line
+        // So now we're actually in the litreal
+        if starting_literal && !line.is_empty() {
+            starting_literal = false;
+            in_literal = true;
+            in_any_code = true;
+            block_indent = line_indent;
+            // TODO: should we not be this aggressive? Let it autodetect?
+            output.push_str("\n```python\n");
+        }
+
+        // If we're not in a codeblock and we see something that signals a doctest, start one
+        if !in_any_code && line.starts_with(">>>") {
+            block_indent = line_indent;
+            in_doctest = true;
+            in_any_code = true;
+            // TODO: is there something more specific? `pycon`?
+            output.push_str("```python\n");
+        }
+
+        // If we're not in a codeblock and we see something that signals a literal block, start one
+        if !in_any_code && let Some(without_lit) = line.strip_suffix("::") {
```

Comment:
I'd have to play with restructuredText but the way I read this paragraph

> Literal code blocks ([ref](https://docutils.sourceforge.io/docs/ref/rst/restructuredtext.html#literal-blocks)) are introduced by ending a paragraph with the special marker ::. The literal block must be indented

is that what comes after is only a literal block if there's at least one blank line, and text is indented. 

I think your code correctly handles the empty line. But it ends up pushing an empty markdown code block if the next paragraph isn't correctly indented.